operator/src/lib.rs

```diff
@@ -572,85 +283,185 @@ impl SparkState {
+        //      - Individual nodes
+        if let Some(nodes_for_role) = self.eligible_nodes.get(&node_type) {
+            for (role_group, nodes) in nodes_for_role {
+                // Create config map for this role group
+                let pod_name = pod_utils::create_pod_name(
+                    &self.context.name(),
+                    role_group,
+                    &node_type.to_string(),
+                );
+
+                let cm_name = create_config_map_name(&pod_name);
+                debug!("pod_name: [{}], cm_name: [{}]", pod_name, cm_name);
+
+                // extract config
+                let config = self.context.resource.spec.get_config(node_type, role_group);
+                self.create_config_map(&cm_name, config).await?;
+
+                debug!(
+                    "Identify missing pods for [{}] role and group [{}]",
+                    node_type, role_group
+                );
+                trace!(
+                    "candidate_nodes[{}]: [{:?}]",
+                    nodes.len(),
+                    nodes
+                        .iter()
+                        .map(|node| node.metadata.name.as_ref().unwrap())
+                        .collect::<Vec<_>>()
+                );
+                trace!(
+                    "existing_pods[{}]: [{:?}]",
+                    &self.existing_pods.len(),
+                    &self
+                        .existing_pods
+                        .iter()
+                        .map(|pod| pod.metadata.name.as_ref().unwrap())
+                        .collect::<Vec<_>>()
+                );
+                trace!(
+                    "labels: [{:?}]",
+                    get_node_and_group_labels(role_group, &node_type)
+                );
+                let nodes_that_need_pods = k8s_utils::find_nodes_that_need_pods(
+                    nodes,
+                    &self.existing_pods,
+                    &get_node_and_group_labels(role_group, &node_type),
+                );
+
+                for node in nodes_that_need_pods {
+                    let node_name = if let Some(node_name) = &node.metadata.name {
+                        node_name
+                    } else {
+                        warn!("No name found in metadata, this should not happen! Skipping node: [{:?}]", node);
+                        continue;
+                    };
+                    debug!(
+                        "Creating pod on node [{}] for [{}] role and group [{}]",
+                        node.metadata
+                            .name
+                            .as_deref()
+                            .unwrap_or("<no node name found>"),
+                        node_type,
+                        role_group
+                    );
+
+                    let master_pods =
+                        filter_pods_for_type(&self.existing_pods, &SparkNodeType::Master);
+
+                    let master_urls =
```

Comment:
This might be executed fairly often, also later when comparing the master url hash..
I might make sense to calculate these once early on and store the urls and hash in the context for later reference.